integrations/github/github.go

```diff
@@ -0,0 +1,436 @@
+				g.clientHasSet.Wait()
+			}
+			g.clientHasSet.L.Unlock()
+		}
+	}
+}
+
+// AttachToApp ...
+func (g *Github) AttachToApp(app core.App) error {
+	g.m.Lock()
+	defer g.m.Unlock()
+	g.init(app)
+
+	appConfig := &githubApp{
+		app: app,
+	}
+	app.Config("github", &appConfig.config)
+	g.apps[app.Name()] = appConfig
+
+	g.setupDeployKey(appConfig)
+	g.setupHooks(appConfig)
+
+	app.Listen(core.SignalBuildStarted, g.onBuildStarted)
+	app.Listen(core.SignalBuildComplete, g.onBuildFinished)
+	return nil
+}
+
+func (g *Github) setupDeployKey(appConfig *githubApp) error {
+	cfg := appConfig.config
+	// TODO - would be nicer to generate ssh key automatically
+	if cfg.PublicKey == "" {
+		logcritf("(%s) No public key available, create one and add it to the configuration", appConfig.app.Name())
+		return errors.New("No pub key available")
+	}
+
+	keyName := fmt.Sprintf("NGBuild ssh deploy key - %s", appConfig.app.Name())
+	_, _, err := g.client.Repositories.CreateKey(cfg.Owner, cfg.Repo, &github.Key{
+		Title:    &keyName,
+		Key:      &cfg.PublicKey,
+		ReadOnly: &[]bool{true}[0],
+	})
+
+	if err != nil && strings.Contains(err.Error(), "key is already in use") == false {
+		logcritf("Couldn't create deploy key for %s: %s", appConfig.app.Name(), err)
+		return err
+	}
+
+	return nil
+}
+
+func (g *Github) setupHooks(appConfig *githubApp) {
+	cfg := appConfig.config
+	_, _, err := g.client.Repositories.Get(cfg.Owner, cfg.Repo)
+	if err != nil {
+		logwarnf("(%s) Repository does not exist, owner=%s, repo=%s", appConfig.app.Name(), cfg.Owner, cfg.Repo)
+		return
+	}
+
+	hookURL := fmt.Sprintf("%s/cb/github/hook/%s", core.GetHTTPServerURL(), appConfig.app.Name())
+	_, _, err = g.client.Repositories.CreateHook(cfg.Owner, cfg.Repo, &github.Hook{
+		Name:   &[]string{"web"}[0],
+		Active: &[]bool{true}[0],
+		Config: map[string]interface{}{
+			"url":          hookURL,
+			"content_type": "json",
+		},
+		Events: []string{"pull_request",
+			"delete",
+			"issue_comment",
+			"pull_request_review",
+			"pull_request_review_event",
+			"push",
+			"status",
+		},
+	})
+	if err != nil && strings.Contains(err.Error(), "Hook already exists") == false {
+		logwarnf("Could not create webhook, owner=%s, repo=%s: %s", cfg.Owner, cfg.Repo, err)
+		return
+	}
+
+}
+
+// Shutdown ...
+func (g *Github) Shutdown() {}
+
+func (g *Github) trackPullRequest(app *githubApp, event *github.PullRequestEvent) {
+	if event.PullRequest == nil {
+		logcritf("pull request is nil")
+		return
+	}
+	pull := event.PullRequest
+	pullID := strconv.Itoa(*pull.ID)
+	g.m.Lock()
+	defer g.m.Unlock()
+
+	// check for ignored branches
+	for _, branchIgnore := range app.config.IgnoredBranches {
+		if branchIgnore == *pull.Base.Ref {
+			logwarnf("Ignoring pull request, is an ignored branch")
+			return
+		}
+	}
+
+	g.trackedPullRequests[pullID] = pullRequestStatus{
+		pull: pull,
+	}
+	g.buildPullRequest(app, pull)
+}
+
+func (g *Github) buildPullRequest(app *githubApp, pull *github.PullRequest) {
```

Comment:
i don't know because github won't let me go to this line in the code or expand this view up or down because github is ridiculous now, but i have changes coming in, in a bit